ast/src/analyzed/mod.rs

```diff
@@ -357,43 +356,38 @@ fn substitute_intermediate<T: Copy>(
 
 /// Recursively inlines intermediate polynomials inside an expression and returns the new expression
 /// This uses a cache to avoid resolving an intermediate polynomial twice
-fn inlined_expression_from_intermediate_poly_id<T: Copy>(
-    poly_id: u64,
-    intermediate_polynomials: &HashMap<u64, &AlgebraicExpression<T>>,
-    cache: &mut HashMap<(u64, bool), AlgebraicExpression<T>>,
-    next: bool,
+///
+/// poly_to_replace can be a "next" reference, but then its value cannot contain any next references.
+fn inlined_expression_from_intermediate_poly_id<T: Copy + Display>(
+    poly_to_replace: AlgebraicReference,
+    intermediate_polynomials: &HashMap<PolyID, &AlgebraicExpression<T>>,
+    cache: &mut HashMap<AlgebraicReference, AlgebraicExpression<T>>,
 ) -> AlgebraicExpression<T> {
-    if let Some(e) = cache.get(&(poly_id, next)) {
+    assert_eq!(poly_to_replace.poly_id.ptype, PolynomialType::Intermediate);
+    if let Some(e) = cache.get(&poly_to_replace) {
         return e.clone();
     }
-    let mut expr = intermediate_polynomials[&poly_id].clone();
+    let mut expr = intermediate_polynomials[&poly_to_replace.poly_id].clone();
     expr.post_visit_expressions_mut(&mut |e| {
-        if let AlgebraicExpression::Reference(r) = e {
-            if next {
-                assert!(!r.next, "Next operator is applied twice!");
-                r.next = true;
-            }
-            match r.poly_id.ptype {
-                PolynomialType::Committed => {}
-                PolynomialType::Constant => {}
-                PolynomialType::Intermediate => {
-                    // read from the cache, if no cache hit, compute the inlined expression
-                    *e = cache
-                        .get(&(r.poly_id.id, r.next))
-                        .cloned()
-                        .unwrap_or_else(|| {
-                            inlined_expression_from_intermediate_poly_id(
-                                r.poly_id.id,
-                                intermediate_polynomials,
-                                cache,
-                                r.next,
-                            )
-                        });
-                }
+        let AlgebraicExpression::Reference(r) = e else { return; };
+        // "forward" the next operator from the polynomial to be replaced.
+        if poly_to_replace.next && r.next {
+            let value = intermediate_polynomials[&poly_to_replace.poly_id];
+            panic!("Error inlining intermediate polynomial {poly_to_replace} = {value}:\nNext operator already applied to {poly_to_replace} but also in {r} - cannot apply it twice!");
```

Comment:
```suggestion
            panic!("Error inlining intermediate polynomial {poly_to_replace} = {value}':\nNext operator already applied to {} and then again to {} - cannot apply it twice!", r.name, poly_to_replace.name);
```

A bit clearer in my opinion. Example:
```
Error inlining intermediate polynomial main_arith.carry' = ((main_arith.carry_high' * 65536) + main_arith.carry_low)':
Next operator already applied to main_arith.carry_high and then again to main_arith.carry - cannot apply it twice!
```

Reply:
true!